trio/_core/_gen_exports.py

```diff
@@ -0,0 +1,147 @@
+import ast
+import astor
+import os
+import textwrap
+
+SOURCE_TREE = './trio/_core'
+
+
+def get_public_methods(tree):
+    """ Return a list of methods marked as public.
+    The function walks the given tree and extracts
+    all objects that are functions and have a
+    doc string that starts with PUBLIC
+    """
+    methods = []
+    for node in ast.walk(tree):
+        for child in ast.iter_child_nodes(node):
+            if (isinstance(child, ast.FunctionDef) or
+               isinstance(child, ast.AsyncFunctionDef)) and \
+               get_doc_string(child).startswith('PUBLIC '):
+                child.parent = node
+                child.module_file = tree.module_file
+                child.module_path = tree.module_path
+                methods.append(child)
+    return methods
+
+
+def get_module_trees_by_dir(source_dir):
+    """ Converts a list of modules into ast module objects
+    """
+    mod_files = astor.code_to_ast.find_py_files(source_dir)
+    modules = []
+    for mod_file in mod_files:
+        if not mod_file[1].startswith('_public'):
+            module = astor.code_to_ast.parse_file(os.path.join(*mod_file))
+            module.module_path = mod_file[0]
+            module.module_file = mod_file[1]
+            modules.append(module)
+    return modules
+
+
+def get_export_modules_by_dir(source_dir):
+    export_modules = []
+    modules = get_module_trees_by_dir(source_dir)
+
+    for module in modules:
+        methods = get_public_methods(module)
+        if methods:
+            export_modules.append(module)
+    return export_modules
+
+
+def get_doc_string(func):
+    """ Returns the doc string of a function
+    or an empty sting if none
+    """
+    doc = func.body[0]
+    if isinstance(doc, ast.Expr):
+        if hasattr(doc, 'value'):
+            if hasattr(doc.value, 's'):
+                return doc.value.s
+    return ""
+
+
+def gen_exports():
+    """ Sync the existing functions with the list
+    of exported functions imported from _public
+    """
+
+    imports = """from ._run import GLOBAL_RUN_CONTEXT, Runner
+from ._ki import LOCALS_KEY_KI_PROTECTION_ENABLED
+
+
+"""
```

Comment:
We should also put a big prominent comment at the top warning that this is an auto-generated file, and should not be edited directly.